### ISSUE TITLE
Request: Photo Comments

### ISSUE THREAD

**Issue by Jwink3101** [RhetTbull/osxphotos#214]
Photos app (at least on Catalina+) stores comments in, say, a shared album. I did not seem to see a way to extract them from osxphotos API. Is there an easy way? I've looked at the database as sqlite and it is a mess but the comments are there!

Any ideas? I am also happy to try to help!

**Comment by RhetTbull** [comment]
Hi @Jwink3101 currently `osxphotos` does not provide a way to get the comments.  A quick investigation indicates these are stored in `ZCLOUDSHAREDCOMMENT` table and are referenced in `ZGENERICASSET`.

This sounds like a useful addition.  I'm currently busy with another project but will add this to the queue as an enhancement.  If you'd like to try decoding the schema for associating comments with items the osxphotos [wiki](https://github.com/RhetTbull/osxphotos/wiki) has good info on the database schema. 

For example:

```sql
CREATE TABLE ZCLOUDSHAREDCOMMENT (
  Z_PK INTEGER PRIMARY KEY, Z_ENT INTEGER, 
  Z_OPT INTEGER, ZISBATCHCOMMENT INTEGER, 
  ZISCAPTION INTEGER, ZISDELETABLE INTEGER, 
  ZISLIKE INTEGER, ZISMYCOMMENT INTEGER, 
  ZCLOUDFEEDCOMMENTENTRY INTEGER, 
  ZCLOUDFEEDLIKECOMMENTENTRY INTEGER, 
  ZCOMMENTEDASSET INTEGER, Z34_COMMENTEDASSET INTEGER, 
  ZLIKEDASSET INTEGER, Z34_LIKEDASSET INTEGER, 
  Z_FOK_LIKEDASSET INTEGER, Z_FOK_CLOUDFEEDCOMMENTENTRY INTEGER, 
  Z_FOK_COMMENTEDASSET INTEGER, Z_FOK_CLOUDFEEDLIKECOMMENTENTRY INTEGER, 
  ZCOMMENTCLIENTDATE TIMESTAMP, ZCOMMENTDATE TIMESTAMP, 
  ZCLOUDGUID VARCHAR, ZCOMMENTTEXT VARCHAR, 
  ZCOMMENTTYPE VARCHAR, ZCOMMENTERHASHEDPERSONID VARCHAR
);
```

These are also referenced in `ZGENERICASSET` (which holds most of the primary details for each item):

```sql
CREATE TABLE ZGENERICASSET (
  Z_PK INTEGER PRIMARY KEY, 
  Z_ENT INTEGER, 
  Z_OPT INTEGER, 
  ZAVALANCHEPICKTYPE INTEGER, 
  ZCLOUDDELETESTATE INTEGER, 
  ZCLOUDDOWNLOADREQUESTS INTEGER, 
  ZCLOUDHASCOMMENTSBYME INTEGER, 
  ZCLOUDHASCOMMENTSCONVERSATION INTEGER, 
  ZCLOUDHASUNSEENCOMMENTS INTEGER, 
  ...
  ZCLOUDLASTVIEWEDCOMMENTDATE TIMESTAMP, 
  ...
```

**Comment by RhetTbull** [comment]
This sql query will associate the shared cloud comments with a photo UUID:

```sql
SELECT 
ZGENERICASSET.ZUUID, --UUID of the photo
ZCLOUDSHAREDCOMMENT.ZISBATCHCOMMENT,  -- ?
ZCLOUDSHAREDCOMMENT.ZISCAPTION, --comment is also photo caption
ZCLOUDSHAREDCOMMENT.ZISDELETABLE, --0 if someone else's comment, 1 if user's own comment
ZCLOUDSHAREDCOMMENT.ZISLIKE, --comment is actually a "like"
ZCLOUDSHAREDCOMMENT.ZISMYCOMMENT, --my (user's) comment
ZCLOUDSHAREDCOMMENT.ZCLOUDFEEDCOMMENTENTRY, -- ??
ZCLOUDSHAREDCOMMENT.ZCLOUDFEEDLIKECOMMENTENTRY, -- ??
ZCLOUDSHAREDCOMMENT.Z34_COMMENTEDASSET, -- ??
ZCLOUDSHAREDCOMMENT.ZLIKEDASSET, --Foreign key to map to Z_PK in ZGENERICASSETS
ZCLOUDSHAREDCOMMENT.ZCOMMENTEDASSET, --Foreign key to map to Z_PK in ZGENERICASSETS
ZCLOUDSHAREDCOMMENT.Z34_LIKEDASSET, -- ??
ZCLOUDSHAREDCOMMENT.Z_FOK_LIKEDASSET, -- sort order for Likes
ZCLOUDSHAREDCOMMENT.Z_FOK_CLOUDFEEDCOMMENTENTRY, -- ??
ZCLOUDSHAREDCOMMENT.Z_FOK_COMMENTEDASSET, -- sort order for comments on a photo
ZCLOUDSHAREDCOMMENT.Z_FOK_CLOUDFEEDLIKECOMMENTENTRY,  -- ??
ZCLOUDSHAREDCOMMENT.ZCOMMENTDATE, -- date of comment
ZCLOUDSHAREDCOMMENT.ZCOMMENTTEXT, -- text of comment
ZCLOUDSHAREDCOMMENT.ZCOMMENTERHASHEDPERSONID -- hashed ID that maps to the person. For person that shared album, the full name can be retrieved from ZGENERICALBUM.ZCLOUDOWNDERFULLNAME by mapping ZCOMMENTERHASHEDPERSONID to ZGENERICALBUM.ZCLOUDOWNERHASHEDPERSONID but not sure about other commenters
FROM
ZCLOUDSHAREDCOMMENT
JOIN ZGENERICASSET ON
ZGENERICASSET.Z_PK = ZCLOUDSHAREDCOMMENT.ZCOMMENTEDASSET
OR
ZGENERICASSET.Z_PK = ZCLOUDSHAREDCOMMENT.ZLIKEDASSET
```

**Comment by Jwink3101** [comment]
Wow! I've been working on it myself but using a mixture of python and SQL. And an SQL dump to quickly grep for where some data is. There is certainly duplicate data on there. 

I am an SQL noob so it's cool to see how it really should be done! I was fine with a Jacky version since multiple queries are fast on SQLite. 

Thanks for sharing. It's always good to be learning new things!

**Comment by RhetTbull** [comment]
I'm no SQL pro and there are a lot more succinct ways to write this query, for example, you could use an [alias](https://www.w3schools.com/sql/sql_alias.asp) to avoid having to write "ZCLOUDSHAREDCOMMENT" over and over. However, I did to be very explicit in all my queries for `osxphotos` because I find it to be more maintainable and more clear to someone browsing the code to understand the Photos database.

I want to eventually write an article for the wiki to document the reverse engineering tools and techniques I've used.  A few things I've found useful:

- Understand the SQL schema, for example, dump it with `sqlite3 Photos.sqlite ".schema"` so you can see the actual tables and names, which for the most part, have useful names
- Use a database browser to look through the database -- I like [DB Browser for SQLite](https://sqlitebrowser.org/) though there are many others.  Note: DB Browser is not a native Mac app and is unable to "view package contents" to see inside the .photoslibrary package.  You need to manually copy the database folder out of the .photoslibrary you're investigating to another location for DB Browser to access it.
- Diff the database. For example, make a copy of the database, do some operation (like a add a comment to a photo), then compare the two.  [sqldiff](https://sqlite.org/sqldiff.html) is good for this.
- Sometimes I just dump the entire database to CSV text files so I can `grep` for text I'm interested in or browser in my editor.  [Here's](https://gist.github.com/RhetTbull/f1de10a519ff193a36251761b6f1e372) a short python script that I use for doing that.

**Comment by RhetTbull** [comment]
@Jwink3101 this might get you started if you just want the comments and likes. I realized in throwing this together, there's a bug in `osxphotos` in that there's no easy way to get a shared album (only shared albums have comments) so I've opened issue #217 for that.

```python
""" get shared comments associated with a photo """

import sys

import osxphotos
from osxphotos.albuminfo import AlbumInfo


def main():
    if len(sys.argv) > 1:
        # library as first argument
        photosdb = osxphotos.PhotosDB(dbfile=sys.argv[1])
    else:
        # open default library
        photosdb = osxphotos.PhotosDB()

    # shared albums
    shared = photosdb._get_album_uuids(shared=True)
    for album_uuid in shared:
        album = AlbumInfo(db=photosdb, uuid=album_uuid)
        print(f"Processing album {album.title}")
        # only shared albums can have comments
        # kludgy but it works -- realized there's no easy way to get shared albums
        for photo in album.photos:
            comments, likes = get_comments(photosdb, photo.uuid)
            if comments or likes:
                print(
                    f"{photo.uuid}, {photo.original_filename}, {likes}, {'; '.join(comments)}"
                )


def get_comments(photosdb, uuid):
    """ return comments and likes, if any, for photo with uuid

    Args:
        photosdb: a osxphotos.PhotosDB object
        uuid: uuid of the photo
    
    Returns:
        tuple of (list of comments or [] if no comments, int number of likes)
    """
    conn, cursor = photosdb.get_db_connection()
    results = conn.execute(
        """
        SELECT 
        ZGENERICASSET.ZUUID, --UUID of the photo
        ZCLOUDSHAREDCOMMENT.ZISLIKE, --comment is actually a "like"
        ZCLOUDSHAREDCOMMENT.Z_FOK_COMMENTEDASSET, -- sort order for comments on a photo
        ZCLOUDSHAREDCOMMENT.ZCOMMENTDATE, -- date of comment
        ZCLOUDSHAREDCOMMENT.ZCOMMENTTEXT -- text of comment
        FROM ZCLOUDSHAREDCOMMENT
        JOIN ZGENERICASSET ON
        ZGENERICASSET.Z_PK = ZCLOUDSHAREDCOMMENT.ZCOMMENTEDASSET
        OR
        ZGENERICASSET.Z_PK = ZCLOUDSHAREDCOMMENT.ZLIKEDASSET
        WHERE ZGENERICASSET.ZUUID = ?
    """,
        ([uuid]),
    ).fetchall()

    comments = []
    likes = 0
    comment_sort_order = []
    for row in results:
        if row[1] == 1:
            # it's a like
            likes += 1
        elif row[4] is not None:
            # comment
            comments.append(row[4])
            comment_sort_order.append(row[2])
    if comments:
        comments = sort_list_by_keys(comments, comment_sort_order)
    return (comments, likes)


def sort_list_by_keys(values, sort_keys):
    """ Sorts list values by a second list sort_keys
        e.g. given ["a","c","b"], [1, 3, 2], returns ["a", "b", "c"]

    Args:
        values: a list of values to be sorted
        sort_keys: a list of keys to sort values by
    
    Returns:
        list of values, sorted by sort_keys
    
    Raises:
        ValueError: raised if len(values) != len(sort_keys)
    """
    if len(values) != len(sort_keys):
        return ValueError("values and sort_keys must have same length")

    return list(zip(*sorted(zip(sort_keys, values))))[1]


if __name__ == "__main__":
    main()
```

**Comment by Jwink3101** [comment]
Hey. Thanks for all of that. Again, it is cool to see how it is being done. I have hacked something together so far that does this. My needs are actually a bit more complicated and i am not opposed to a hack (my rule is that the process to get there can be hacky but the result needs to be clear since I may not be the consumer).

I am currently manually downloading the photos from the shared album (which have a different naming scheme than in the database and doesn't seem stored in it based on grepping the whole SQL dump). But I will likely move to an export from your tool.

I am them using a few [image hashes](https://github.com/JohannesBuchner/imagehash) to find duplicates in my own library (since my wife or I do most of the posting). This way I can map the comments back. It kind of sucks that Shared Albums do not store the original but I digress. I also have to careful with how I search since it can easily be O(N^2) but I've found I can converts the hashes to NumPy arrays and use some `scipy.distance` tools to speed it up at the cost of ram.

I have been using those tools you mention. I use the DB Browser and I just did a full sql dump (I thought about individual tables to CSV but I am mostly fine with the browser tool)

**Comment by Jwink3101** [comment]
Also worth noting that if this is going to go into your main code, it is also good to get the persons name. It was in one of your queries above but not the example python code.

Also, are you able to ascertain the timezone of these comments? I am not seeing anything like it in the db but I am also still exploring

**Comment by RhetTbull** [comment]
I wasn't sure at first I could get the name of commenter but I think I can extract if from `ZCLOUDSHAREDALBUMINVITATIONRECORD` which shows who's been invited to a shared album.  The `ZINVITEEHASHEDPERSONID` column matches the `ZCLOUDSHAREDCOMMENT.ZCOMMENTERHASHEDPERSONID` and can be used to lookup `ZINVITEEFULLNAME`.

```sql
CREATE TABLE ZCLOUDSHAREDALBUMINVITATIONRECORD (
  Z_PK INTEGER PRIMARY KEY, Z_ENT INTEGER, 
  Z_OPT INTEGER, ZINVITATIONSTATE INTEGER, 
  ZINVITATIONSTATELOCAL INTEGER, ZINVITEEEMAILKEY INTEGER, 
  ZISMINE INTEGER, ZALBUM INTEGER, Z_FOK_ALBUM INTEGER, 
  ZINVITEESUBSCRIPTIONDATE TIMESTAMP, 
  ZALBUMGUID VARCHAR, ZCLOUDGUID VARCHAR, 
  ZINVITEEFIRSTNAME VARCHAR, ZINVITEEFULLNAME VARCHAR, 
  ZINVITEEHASHEDPERSONID VARCHAR, 
  ZINVITEELASTNAME VARCHAR
```

As for timezone of the comment, as best as I can tell it's in local timezone (timezone naive).  To get the actual date, you need to add a delta to convert from unix epoch to Apple epoch (number of seconds between 1 Jan 1970 and 1 Jan 2001).

Here's an example where the value for `ZCOMMENTDATE = 622428072.180951`

```
>>> import datetime
>>> from datetime import datetime
>>> TIME_DELTA = (datetime(2001, 1, 1, 0, 0) - datetime(1970, 1, 1, 0, 0)).total_seconds()
>>> comment_date = datetime.fromtimestamp(622428072.180951 + TIME_DELTA)
>>> comment_date
datetime.datetime(2020, 9, 21, 20, 41, 12, 180951)
```

**Comment by RhetTbull** [comment]
@Jwink3101 I've added a more comprehensive [example](https://github.com/RhetTbull/osxphotos/blob/master/examples/get_shared_photo_comments.py) to the `examples` folder that gets comments and likes along with user name and date for photos.  This will serve as the basis for future incorporation into `osxphotos` but for now should provide an easy way to get a list of comments/likes for a given photo (sorted by date).  The comments/likes are returned as [dataclass](https://docs.python.org/3/library/dataclasses.html) objects for easy manipulation.

**Comment by Jwink3101** [comment]
Thanks for the update. I ended up reversing it all myself with some hints from your tools and this discussion. I am still debugging getting the actual export of the photos to work (currently doing it manually within Photos) but I have not ruled out user error. When/if I do, I will create its own ticket.

Also may try [GrannySmith](https://tyler.io/grannysmith/)

**Comment by RhetTbull** [comment]
Comments/Likes have been implemented in release [v0.36.0](https://github.com/RhetTbull/osxphotos/releases/tag/v0.36.0)
